core/src/main/java/org/jdbi/v3/core/mapper/reflect/internal/ImmutablesPropertiesFactory.java

```diff
@@ -0,0 +1,266 @@
+            super(type);
+            this.defn = defn;
+            this.builder = builder;
+            for (Method m : defn.getMethods()) {
+                if (isProperty(m)) {
+                    final String name = m.getName();
+                    properties.put(name, createProperty(name, m));
+                }
+            }
+        }
+
+        private boolean isProperty(Method m) {
+            return m.getParameterCount() == 0
+                && !m.isSynthetic()
+                && !Modifier.isStatic(m.getModifiers())
+                && m.getDeclaringClass() != Object.class;
+        }
+
+        @Override
+        public Map<String, ? extends PojoProperty<T>> getProperties() {
+            return properties;
+        }
+
+        protected abstract ImmutablesPojoProperty<T> createProperty(String name, Method m);
+    }
+
+    static class ImmutablePojoProperties<T, B> extends BasePojoProperties<T, B> {
+        private MethodHandle builderBuild;
+
+        ImmutablePojoProperties(Type type, Class<?> defn, Supplier<B> builder) {
+            super(type, defn, builder);
+            try {
+                builderBuild = MethodHandles.lookup().unreflect(builder.get().getClass().getMethod("build"));
+            } catch (NoSuchMethodException | IllegalAccessException e) {
+                throw new IllegalArgumentException("Failed to inspect Immutables " + defn, e);
+            }
+        }
+        @Override
+        protected ImmutablesPojoProperty<T> createProperty(String name, Method m) {
+            final Class<?> builderClass = builder.get().getClass();
+            try {
+                final Type propertyType = GenericTypes.resolveType(m.getGenericReturnType(), getType());
+                return new ImmutablesPojoProperty<T>(
+                        name,
+                        QualifiedType.of(propertyType, Qualifiers.getQualifiers(m)),
+                        m,
+                        alwaysSet(),
+                        MethodHandles.lookup().unreflect(m),
+                        findBuilderSetter(builderClass, name, propertyType));
+            } catch (IllegalAccessException | NoSuchMethodException e) {
+                throw new IllegalArgumentException("Failed to inspect method " + m, e);
+            }
+        }
+
+        private MethodHandle findBuilderSetter(final Class<?> builderClass, String name, Type type)
+        throws IllegalAccessException, NoSuchMethodException {
+            final NoSuchMethodException failure;
+            try {
+                return MethodHandles.lookup().unreflect(builderClass.getMethod(name, GenericTypes.getErasedType(type)));
+            } catch (NoSuchMethodException e) {
+                failure = e;
+            }
+            for (Method m : builderClass.getMethods()) {
+                if (m.getName().equals(name) && m.getParameterCount() == 1) {
+                    return MethodHandles.lookup().unreflect(m);
```

Comment:
So if we don't find the method with the exact parameter type, search for single-parameter method with the right name and hope the parameter type is right?